pkg/controllers/cronfederatedhpa/cronfederatedhpa_job.go

```diff
@@ -0,0 +1,280 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cronfederatedhpa
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gorhill/cronexpr"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	autoscalingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/autoscaling/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+)
+
+type CronFederatedHPAJob struct {
+	client        client.Client
+	eventRecorder record.EventRecorder
+
+	namespaceName types.NamespacedName
+	rule          autoscalingv1alpha1.CronFederatedHPARule
+}
+
+func NewCronFederatedHPAJob(client client.Client, eventRecorder record.EventRecorder,
+	namespaceName types.NamespacedName, rule autoscalingv1alpha1.CronFederatedHPARule) *CronFederatedHPAJob {
+	return &CronFederatedHPAJob{
+		client:        client,
+		eventRecorder: eventRecorder,
+		namespaceName: namespaceName,
+		rule:          rule,
+	}
+}
+
+func (c *CronFederatedHPAJob) Run() {
+	klog.V(4).Infof("Start to handle CronFederatedHPA %s", c.namespaceName)
+	defer klog.V(4).Infof("End to handle CronFederatedHPA %s", c.namespaceName)
+
+	var err error
+	cronFHPA := &autoscalingv1alpha1.CronFederatedHPA{}
+	err = c.client.Get(context.TODO(), c.namespaceName, cronFHPA)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("CronFederatedHPA %s not found", c.namespaceName)
+		} else {
+			klog.Errorf("Get CronFederatedHPA failed: %v", err)
+		}
+		return
+	}
+
+	// It shouldn't be true, because we will not add the rule to schedule queue
+	if helper.IsCronFederatedHPARuleSuspend(c.rule) {
+		klog.V(4).Infof("CronFederatedHPA %s Rule %s is suspended", c.namespaceName, c.rule.Name)
+		return
+	}
+
+	var scaleErr error
+	defer func() {
+		if scaleErr != nil {
+			c.eventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ScaleFailed", scaleErr.Error())
+			err = c.addFailedExecutionHistory(cronFHPA, scaleErr.Error())
+		} else {
+			err = c.addSuccessExecutionHistory(cronFHPA, c.rule.TargetReplicas, c.rule.TargetMinReplicas, c.rule.TargetMaxReplicas)
+		}
+		if err != nil {
+			c.eventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ScaleFailed", err.Error())
```

Comment:
It is best to differentiate the `Event` here from the `Event` in L84, as it is possible that `Scale` to the workload was successful, only the `status` update failed.